packaging/windows/installer.nsi

```diff
@@ -205,9 +194,51 @@ Function .onInit
       ; At this point, I'm very much puzzled as why writing NSIS installer
       ; feels like reverse engineering a taiwanese NES clone...
       ExecWait '"$R0" /S _?=$R1'
+FunctionEnd
 
-    done:
+Function checkUninstaller
+  ; Check either 32-bit or 64-bit registry for parsec uninstaller
+  ReadRegStr $R0 ${PROGRAM_UNINST_ROOT_KEY} \
+    "${PROGRAM_UNINST_KEY}" \
+    "UninstallString"
+    StrCmp $R0 "" done
 
+  ; Check that the uninstaller file actually exists
+  IfFileExists $R0 ask_for_uninstall cleanup
+
+  ; Perform some registry cleanup if the installer wasn't found
+  cleanup:
+    DeleteRegKey ${PROGRAM_UNINST_ROOT_KEY} "${PROGRAM_UNINST_KEY}"
+    Goto done
+
+  ; Prompt the user for uninstaller the previous version of parsec
+  ask_for_uninstall:
+    MessageBox MB_OKCANCEL|MB_ICONEXCLAMATION \
+      "${PROGRAM_NAME} is already installed. $\n$\nClick `OK` to remove the \
+      previous version or `Cancel` to cancel this upgrade." \
+      /SD IDOK IDOK uninstall IDCANCEL _abort
+
+  ; Abort in case the user clicked "cancel"
+  _abort:
+    Abort
+
+  ; Run the uninstaller
+  uninstall:
+    Call runUninstaller
+    Goto done
```

Comment:
```suggestion
  uninstall:
    Call runUninstaller
    Call cleanup
    Goto done
```

Once the `runUninstaller` routine has finished, should we cleanup the register values ?

Reply:
The uninstaller is meant to cleanup those values, so I chose to cleanup the registry only if the uninstaller no longer exists. It's kind of arbitrary tho

Reply:
Ok, so as you said, the uninstaller remove the register value then the cleanup just after is redundant